### PR TITLE
perf: expand torch.compile coverage to 4bit dequant

### PR DESCRIPTION
**Pull request by thad0ctor**
# Description

Expands torch compile coverage for 4bit lora fine tuning by registering axolotl's NF4 dequant fast path as a `torch.library.custom_op` (`torch.ops.axolotl.nf4_dequantize`) with `register_fake`. `dequantize()` branches on `torch.compiler.is_compiling()`: eager calls the ctypes body directly (zero overhead), tracing dispatches through the opaque op so Dynamo can compile around it. Also drops dead legacy-list `quant_state` branch, unused `out=` parameter, and the `HAS_CUDA_STREAM` version gate. Test fixtures rewritten to use real `bnb.functional.quantize_4bit` artifacts (the old synthetic fixtures used `blocksize=32`, which is bnb-invalid, the old ctypes path silently skipped validation) following recent bnb PR (https://github.com/bitsandbytes-foundation/bitsandbytes/pull/1544).

## Motivation and Context

`torch_compile: true` on QLoRA appears to "work" on current bnb (≥0.46 ships registered ops) but Dynamo still encounters the Unsloth-derived axolotl fast path's raw ctypes calls. It can't trace `ctypes.c_int(...)` (`_SimpleCData.__new__`) or the FFI symbols, and ends up recompiling `bitsandbytes.functional.get_ptr` against different tensor dtypes (Byte for packed weight, Float for absmax) until it hits `recompile_limit (8)` and falls back to eager for that frame, resulting is silent compile inefficiency.

Wrapping the fast path in a Dynamo-opaque custom op removes both the ctypes trace failure and the `get_ptr` recompile thrash. Eager path is byte-identical to main.

## How has this been tested?

**Unit tests** (`tests/e2e/kernels/test_quantize.py`, 6 cases): all pass on torch 2.11.0 + bnb 0.49.1. Cases cover null state, shape preservation, non-square transposed input with value comparison vs bnb.functional.dequantize_4bit(...).t()`, non-nested (single-quant) fallback, and `torch.compile` correctness on both nested and non-nested paths.

**Test environment:**
- GPU: NVIDIA RTX PRO 6000 Blackwell Workstation (96 GB)
- torch 2.11.0+cu130, bitsandbytes 0.49.1, transformers 5.5.4, Linux 6.14, CUDA 13.0
- Model: Qwen3.5-0.8B
- Config: QLoRA, `lora_mlp_kernel`/`lora_qkv_kernel`/`lora_o_kernel: true`, `micro_batch=2`, `seq_len=512`

**Kernel-level dequant bench (eager), main vs branch, 1000 iters per shape:**

| Shape | main median | branch median | Δ |
|---|---|---|---|
| Qwen 1024² | 31.7 µs | 31.6 µs | -0.3% |
| Qwen 3072×1024 | 34.1 µs | 33.5 µs | -1.7% |
| Llama8B 4096² | 44.8 µs | 44.2 µs | -1.4% |
| Llama8B 14336×4096 | 133.5 µs | 133.2 µs | -0.3% |
| Llama8B 4096×14336 | 135.1 µs | 134.0 µs | -0.8% |
| Llama70B 8192² | 151.4 µs | 150.7 µs | -0.5% |
| Llama70B 8192×28672 | 485.3 µs | 485.3 µs | 0.0% |

Eager is at parity across the matrix (all within ±2% run-to-run noise).

**End-to-end QLoRA training (300 steps, two reversed-order rounds):**

  Training config (/tmp/axolotl_qwen_test/config.yaml):
```
  base_model: <Qwen3.5-0.8B>
  load_in_4bit: true
  adapter: qlora

  # axolotl LoRA fast paths ON — these are what call our dequantize() via autograd.Function
  lora_mlp_kernel: true
  lora_qkv_kernel: true
  lora_o_kernel: true
  
  lora_r: 8
  lora_alpha: 16
  lora_dropout: 0
  lora_target_modules:
    - q_proj, k_proj, v_proj, o_proj
    - gate_proj, up_proj, down_proj
   
  sequence_len: 512
  sample_packing: false
  pad_to_sequence_len: true
  micro_batch_size: 2
  gradient_accumulation_steps: 1
  max_steps: 300

  optimizer: adamw_torch
  lr_scheduler: constant
  learning_rate: 0.0002
  warmup_steps: 0

  bf16: auto
  gradient_checkpointing: false
  flash_attention: false
  attn_implementation: eager
  saves_per_epoch: 0
```

| Variant | r1 wall | r2 wall | avg sec/step | vs main+compile |
|---|---|---|---|---|
| main + eager | 948 s | 705 s | 2.755 | — |
| main + compile | 737 s | 755 s | 2.487 | baseline |
| branch + eager | 795 s | 686 s | 2.468 | — |
| **branch + compile** | **603 s** | **558 s** | **1.935** | **−22.2%** |

Branch + `torch_compile: true` is fastest in both rounds, including the round where it ran first (cold caches). Equivalent to ≈ **+28% throughput** in steps/sec vs main + compile.

**Mechanism validation** (3-step diagnostic with relevant Dynamo warnings counted):

| Dynamo event | main | branch |
|---|---|---|
| `_SimpleCData.__new__` (ctypes) trace failure | 1 | 0 |
| `recompile_limit (8)` hit on `bnb.functional.get_ptr` | 1 | 0 |
| Recompile reason | `tensor 'A' dtype mismatch. expected Float, actual Byte` | (none) |
| Throughput at step 3 | 242 tok/s/gpu | 315 tok/s/gpu |

**End-to-end LoRA pipeline:** trained multiple steps on Qwen3.5-0.8B with axolotl LoRA fast-path kernels enabled (which exercise `dequantize` in both forward and backward via `autograd.Function`), then `axolotl merge-lora`, then loaded the merged model and generated coherent output. Pipeline clean.

**Risk validation matrix** (4-step training runs on physical RTX 3090s, confirming no regression across architectural / config dimensions):

| # | Scenario | Result | Notes |
|---|---|---|---|
| 1 | fp16 compute dtype (Qwen3.5-0.8B QLoRA) | ✅ 4 steps, no crash | exercises `if dtype == torch.float16` branch in `_ctypes_nf4_dequant` |
| 2 | DoRA adapter (Qwen3.5-0.8B, `peft_use_dora: true`) | ✅ 4 steps, losses 1.74→0.61 | hits `kernels/dora.py:122` `dequantize()` call |
| 3 | Different architecture + larger model (Qwen3-Embedding-8B, `Qwen3ForCausalLM` — standard attention, 8B params) | ✅ 4 steps, 11 GiB / 24 GiB on RTX 3090 | confirms dequant works on non-Qwen3.5 architectures and at 8B scale |
| 4 | FSDP 2-rank multi-GPU (Qwen3.5-0.8B on 2× RTX 3090) | ✅ 4 steps, losses 1.74→1.27, per-rank mem logging cross-reduces cleanly | custom op `axolotl::nf4_dequantize` registers idempotently across ranks; per-process `CUDA_STREAM` cache works correctly |

All four scenarios train to completion, exercise `dequantize` in forward + backward, and report sane losses / finite grad norms.

## AI Usage Disclaimer

Yes — Claude Code 4.7 assisted

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Performance improvement

## Summary by CodeRabbit

* **Bug Fixes**
  * Optimized dequantization performance with improved kernel handling for edge cases.

* **Refactor**
  * Simplified dequantization function API by removing legacy parameters and streamlining the interface.

* **Tests**
  * Enhanced test suite to use real quantization outputs instead of mock data, improving coverage validation.

[Review Change Stack](https://app.coderabbit.ai/change-stack/axolotl-ai-cloud/axolotl/pull/3677?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/axolotl/kernels/quantize.py [hunk 1]
```diff
@@ -58,145 +129,39 @@ def dequantize_fp8(
...
-    # Choose appropriate dequantization function
-    fx = (
-        cdequantize_blockwise_fp16_nf4
-        if dtype == torch.float16
-        else cdequantize_blockwise_bf16_nf4
+    return _ctypes_nf4_dequant(
+        W, absmax, code2, absmax2, offset,
+        quant_state.blocksize, state2.blocksize,
+        quant_state.shape, quant_state.dtype,
     )
```
[the remark this answers is not in the document]
Comment (thad0ctor) [reply src/axolotl/kernels/quantize.py:180]: addressed in PR [e0c629d](https://github.com/axolotl-ai-cloud/axolotl/pull/3677/commits/e0c629db1087a2a01b95738305962f6d8c86e128)

**Review by winglian — APPROVED** [review]
